active/0000-json-patch-support-for-ember-data.md

```diff
@@ -0,0 +1,114 @@
+
+* When records are related with an 'async' option; in order to serialize the record with an acurate list of related ids, every related record's promise must be resolved first. Instead of using Ember Data relationships, perhaps using computed properties would provide synchronous access to the list of related ids when serializing.
+
+* When the RESTAdapter makes a PUT request and the server returns a payload the serializer's extractSingle or extractArray which normally expects the complete payload, not a partial, see [DS.RESTAdapter.html#ajax]. Perhaps a 206 (partial) repsonse can be supported using [DS.Store.html#update] to update the related records.
+
+Each one of these cases may involve creating, updating or deleting using RESTful requests. It may be possible for a server to return a payload that has changes to be applied to the application's store (distributed cache, in memory). Some of the cases above may include implementations on both the client and server. Perhaps server responses can be tuned to supply the necessary changes after PUT and DELETE request complete.
+
+[bucket]: https://github.com/pixelhandler/ember-bucket
+[Ember.DAG]: https://github.com/emberjs/ember.js/blob/master/packages/ember-application/lib/system/dag.js
+[Directed acyclic graph]: https://en.wikipedia.org/wiki/Directed_acyclic_graph
+[Ember.Application#initializer]: http://emberjs.com/api/classes/Ember.Application.html#method_initializer
+[DS.Store.html#update]: http://emberjs.com/api/data/classes/DS.Store.html#method_update
+[DS.RESTAdapter.html#ajax]: http://emberjs.com/api/data/classes/DS.RESTAdapter.html#method_ajax
+
+
+# Detailed design
+
+Create functionality that supports [JSON Patch] operations. Using the JSON Patch specification, [rfc6902], for change operations, a model can manage updates on a per property basis using (HTTP) PATCH requests. 
+
+Instead of sending a request with a complete document, a patch operation can be used to persist individual property changes. Upon success of the patch operation, which includes a test to verify the record is not stale, the record may transition to a clean state.  The model's state machine may be augmented to add isPatching[Property] sub-states that can be used to disable web form inputs or other action buttons; or perhaps using flags for each field that represent the state of that field, active or disabled.
+
+```json
+[
+  { "op": "replace", "path": "/resource/id/property", "value": 42 },
+  { "op": "test", "path": "/resource/id/property", "value": 0 }
+]
+```
+
+[JSON Patch]: http://jsonpatch.com
+[rfc6902]: http://tools.ietf.org/html/rfc6902
+
+## Technical Considerations
+
+* Mixin for (Ember Data) DS.Adapter/DS.Serializer for PATCH requests
+* Mixin for DS.Model to support changes that fire PATCH requests
```

Comment:
@kurko I understand your concern. My intention is that a model can trigger an event on the store service such as `attributeChanged` then the service/adapter can do what it wants, e.g. perhaps call a `patch` method. This is an example of one way that I've implemented this behavior to patch a request as close to real time as possible - https://github.com/pixelhandler/ember-jsonapi-resources/blob/master/README.md#resource-services and here - https://github.com/pixelhandler/ember-jsonapi-resources/wiki/Services#collaborators

I really enjoyed this behavior when I tried out ember-orbit with json-patch. So I've borrowed that behavior for my own persistence needs as well.